lightning-invoice/src/payment.rs

```diff
@@ -399,20 +437,25 @@ where
 	fn retry_payment(
 		&self, payment_id: PaymentId, payment_hash: PaymentHash, params: &RouteParameters
 	) -> Result<(), ()> {
-		let max_payment_attempts = self.retry_attempts.0 + 1;
-		let attempts = *self.payment_cache.lock().unwrap()
+    let attempts = *self.payment_cache.lock().unwrap()
 			.entry(payment_hash)
-			.and_modify(|attempts| *attempts += 1)
-			.or_insert(1);
-
-		if attempts >= max_payment_attempts {
-			log_trace!(self.logger, "Payment {} exceeded maximum attempts; not retrying (attempts: {})", log_bytes!(payment_hash.0), attempts);
-			return Err(());
-		}
+			.and_modify(|attempts| {
+          attempts.counts += 1;
+          attempts.latest_attempted_at = SystemTime::now();
+      })
+      .or_insert(PaymentAttempts::new());
+
+    if ! self.retry_attempts.is_retryable_now(&PaymentAttempts{
+        counts: attempts.counts -1,
```

Comment:
Not quite sure I follow why the count is incremented above and then decremented here.

Reply:
Because the original logic in the code is always add one count(no matter if it's retryable[here](https://github.com/lightningdevkit/rust-lightning/pull/1418/files#diff-5165a75ba263a549c7fb9f4b555e449e44b0b680a80a9715a99d11157802a0d7L405). But since now the retryable logic is encapsulated into a function that expect to take a PaymentAttempts and if we simply passed the updated one, then that will be incorrect(as we want to test if it can be tested before adding 1 count).[](url)

Reply:
Couldn't this be accounted for in `is_retryable_now`? It's adding `1` back as it is.

Reply:
I would say this is _close_ to violating the "no magic numbers" rule, so just something to be mindful of. And by magic numbers i'm referring to this definition: 

> Unique values with unexplained meaning or multiple occurrences which could (preferably) be replaced with named constants

Reply:
So the logic in `retry_payment` is 

1. always add 1 to `counts` in `PaymentAttempts`
2. check if it's retryable
3. the rest of logic

`Retry::is_retryable_now` take `PaymentAttempts` struct and use the counts field to determine if it's retryable if retry strategy is count-based. So in step2, if we don't decrement it by 1, it will be incorrect.


`Retry::Attempts(max_retry_count) => max_retry_count + 1 > attempts.counts,`  the `+ ` is actually correct. For example, 
if `max_retry_count` is 1 and `attempts.count` is 1, this means current retry count is 0(the first attempt is not retry attempt) so the `+ 1` here is required to make this function return true in this case.